Source/Particles/Pusher/ImplicitPushPX.cpp

```diff
@@ -244,13 +243,121 @@ namespace {
+    int *nsuborbits = (HasiAttrib("nsuborbits") ? pti.GetiAttribs("nsuborbits").dataPtr() : nullptr);
+
+    amrex::ParallelFor(
+        np_to_push, [=] AMREX_GPU_DEVICE (long ip)
+    {
+
+        if (nsuborbits && nsuborbits[ip] > 1) {
+            // write signaling flag: how many particles did not converge?
+            amrex::Gpu::Atomic::Add(unconverged_particles_ptr, amrex::Long(1));
+            return;
+        }
+
+    });
+
+    // Setup for handling the suborbit particles. A list of their indices is
+    // gathered, their weights saved, and their weight set to zero (so they
+    // don't contribute to the current density).
+    num_unconverged_particles = *(unconverged_particles.copyToHost());
+    SetupSuborbitParticles(pti, offset, np_to_push, num_unconverged_particles,
+                           unconverged_indices, saved_weights);
+
+}
+
+/*
+ * \brief Setup for handling the suborbit particles. A list of their indices is
+ *        gathered, their weights saved, and their weight set to zero (so they
+ *        don't contribute to the current density).
+ *
+ * \param[in] pti                        The WarpXParIter holding the particles to push
+ * \param[in] offset                     The particle index offset for the particles to be pushed
+ * \param[in] np_to_push                 The number of particles to push
+ * \param[in] num_unconverged_particles  Number of unconverged particles
+ * \param[in/out] unconverged_indices    The list of indices of unconverged particles
+ * \param[in/out] saved_weights          The saved weights of the unconverged particles
+ */
+void
+PhysicalParticleContainer::SetupSuborbitParticles (WarpXParIter & pti,
+                                                   long offset,
+                                                   long np_to_push,
+                                                   long num_unconverged_particles,
+                                                   amrex::Gpu::DeviceVector<long> & unconverged_indices,
+                                                   amrex::Gpu::DeviceVector<amrex::ParticleReal> & saved_weights)
+{
+    // If no unconverged particles, do not do anything
+    if (num_unconverged_particles == 0) { return; }
+
+    auto& attribs = pti.GetAttribs();
+    amrex::ParticleReal* const AMREX_RESTRICT w  = attribs[PIdx::w ].dataPtr() + offset;
+
+    int *nsuborbits = (HasiAttrib("nsuborbits") ? pti.GetiAttribs("nsuborbits").dataPtr() : nullptr);
+
+    auto num_previous = unconverged_indices.size();
+    unconverged_indices.resize(num_previous + num_unconverged_particles);
+    saved_weights.resize(num_previous + num_unconverged_particles);
+
+    long * unconverged_i = unconverged_indices.data() + num_previous;
+    amrex::ParticleReal * saved_w = saved_weights.data() + num_previous;
+
+    if (nsuborbits) {
+        // This looks for the unconverged particles, which had been flagged as invalid
```

Comment:
This comment could be updated since the invalid flag is no longer used.

```suggestion
        // This looks for the particles that require suborbits to converge.
```